q2cli/core/usage.py

```diff
@@ -0,0 +1,156 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2020, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from qiime2.core.type.primitive import Bool
+
+import qiime2.sdk.usage as usage
+from qiime2 import Metadata, MetadataColumn
+from qiime2.sdk.util import (
+    is_metadata_type,
+    is_metadata_column_type,
+    is_visualization_type,
+)
+
+from q2cli.util import to_cli_name
+from q2cli.util import to_snake_case
+
+
+class CLIUsage(usage.Usage):
+    def __init__(self):
+        super().__init__()
+        self._recorder = []
+        self._init_data_refs = dict()
+
+    def _init_data_(self, ref, factory):
+        self._init_data_refs[ref] = factory
+        data = factory()
+        if isinstance(data, MetadataColumn):
+            col = data.name
+            md_file = ref  # Using ref just to make it easy atm
+            return col, md_file
+        return ref
+
+    def _merge_metadata_(self, ref, records):
+        merge_target = ref
+        return merge_target
+
+    def _get_metadata_column_(self, ref, record, column_name):
+        md = record.result
+        return column_name, md
+
+    def _comment_(self, text: str):
+        self._recorder.append('# %s' % (text,))
+
+    def _action_(
+        self, action: usage.UsageAction, input_opts: dict, output_opts: dict
+    ):
+        t = self._template_action(action, input_opts, output_opts)
+        self._recorder.append(t)
+        return output_opts
+
+    def _assert_has_line_matching_(self, ref, label, path, expression):
+        pass
+
+    def render(self):
+        return '\n'.join(self._recorder)
+
+    def get_example_data(self):
+        return {r: f() for r, f in self._init_data_refs.items()}
+
+    def _extract_from_signature(self, action_sig):
+        params, mds = [], []
+        for i, spec in action_sig.parameters.items():
+            q_type = spec.qiime_type
+            if is_metadata_type(q_type):
+                mds.append((i, spec))
+            else:
+                params.append((i, spec))
+        return params, mds
+
+    def _template_action(self, action, input_opts, outputs):
+        action_f, action_sig = action.get_action()
+        cmd = to_cli_name(f"qiime {action_f.plugin_id} {action_f.id}")
+        params, mds = self._extract_from_signature(action_sig)
+        inputs_t = self._template_inputs(action_sig, input_opts)
+        params_t = self._template_parameters(params, input_opts)
+        mds_t = self._template_metadata(mds, input_opts)
+        outputs_t = self._template_outputs(action_sig, outputs)
+        t = " \\\n".join([cmd] + inputs_t + params_t + mds_t + outputs_t)
```

Comment:
Restating what I mentioned above in this review - this line could be refactored as another helper method - it takes the raw command elements and returns an indented, escaped command.